### DISCUSSION TITLE
Play/Pause event in A2DP source mode

### DISCUSSION THREAD

**Discussion by tueddy** [pschatzmann/ESP32-A2DP#299] [Q&A] [answered]
Hi,

i have a connected a kids headphone and can play audio from SD successfully. This is my minimal sample:
https://github.com/tueddy/BTSourceSD.git

Now the headphone has a button for play/pause. Is it possible to get an event so that i can react on it for pausing/playing audio?
I know the other way around using BT-sink and AVRC events which work fine. I wonder if AVRC events for BT-sink only?

Found nothing about in wiki/issues
Thanks in advance
Dirk

**Comment by pschatzmann** [comment]
Just set the log level to debug to see what method gets called when you trigger the events and then you can create your subclass and overwrite the method as explained in https://github.com/pschatzmann/ESP32-A2DP/wiki/Design-Overview

**Reply by tueddy** [reply]
Hi,

i have enabled the verbose level but cannot see any event that track the play/pause button on the BT-headset. Tried with Arduino 2.0.4 / PlatformIO package 5.1.1.

This is my log:

[  9988][D][BluetoothA2DPSource.cpp:455] bt_app_gap_callback(): [BT_AV] bt_app_gap_callback evt 13
[  9989][I][BluetoothA2DPSource.cpp:519] bt_app_gap_callback(): [BT_AV] ESP_BT_GAP_MODE_CHG_EVT mode:2
[ 11215][D][BluetoothA2DPSource.cpp:250] bt_app_work_dispatch(): [BT_API] bt_app_work_dispatch event 0xff00, param len 0
[ 11215][D][BluetoothA2DPSource.cpp:301] bt_app_task_handler(): [BT_API] bt_app_task_handler, sig 0x1, 0xff00
[ 11224][I][BluetoothA2DPSource.cpp:628] bt_app_av_sm_hdlr(): [BT_AV] bt_app_av_sm_hdlr state 5, evt 0xff00
[ 11233][D][BluetoothA2DPSource.cpp:596] process_user_state_callbacks(): [BT_AV] process_user_state_callbacks
[ 11243][D][BluetoothA2DPSource.cpp:730] bt_app_av_state_connected(): [BT_AV] bt_app_av_state_connected evt 65280
[ 11253][D][BluetoothA2DPSource.cpp:790] bt_app_av_media_proc(): [BT_AV] bt_app_av_media_proc evt 65280
[ 11262][I][BluetoothA2DPSource.cpp:795] bt_app_av_media_proc(): [BT_AV] a2dp media ready checking ...
[ 11271][D][BluetoothA2DPSource.cpp:250] bt_app_work_dispatch(): [BT_API] bt_app_work_dispatch event 0x3, param len 16
[ 11282][D][BluetoothA2DPSource.cpp:301] bt_app_task_handler(): [BT_API] bt_app_task_handler, sig 0x1, 0x3
[ 11291][I][BluetoothA2DPSource.cpp:628] bt_app_av_sm_hdlr(): [BT_AV] bt_app_av_sm_hdlr state 5, evt 0x3
[ 11300][D][BluetoothA2DPSource.cpp:596] process_user_state_callbacks(): [BT_AV] process_user_state_callbacks
[ 11310][D][BluetoothA2DPSource.cpp:730] bt_app_av_state_connected(): [BT_AV] bt_app_av_state_connected evt 3
[ 11319][D][BluetoothA2DPSource.cpp:790] bt_app_av_media_proc(): [BT_AV] bt_app_av_media_proc evt 3
[ 11328][I][BluetoothA2DPSource.cpp:801] bt_app_av_media_proc(): [BT_AV] a2dp media ready, starting ...
[ 11444][D][BluetoothA2DPSource.cpp:455] bt_app_gap_callback(): [BT_AV] bt_app_gap_callback evt 13
[ 11445][I][BluetoothA2DPSource.cpp:519] bt_app_gap_callback(): [BT_AV] ESP_BT_GAP_MODE_CHG_EVT mode:0
[ 11477][D][BluetoothA2DPSource.cpp:250] bt_app_work_dispatch(): [BT_API] bt_app_work_dispatch event 0x3, param len 16
[ 11477][D][BluetoothA2DPSource.cpp:301] bt_app_task_handler(): [BT_API] bt_app_task_handler, sig 0x1, 0x3
[ 11478][D][BluetoothA2DPSource.cpp:250] bt_app_work_dispatch(): [BT_API] bt_app_work_dispatch event 0x1, param len 16
[ 11486][I][BluetoothA2DPSource.cpp:628] bt_app_av_sm_hdlr(): [BT_AV] bt_app_av_sm_hdlr state 5, evt 0x3
[ 11505][D][BluetoothA2DPSource.cpp:596] process_user_state_callbacks(): [BT_AV] process_user_state_callbacks
[ 11515][D][BluetoothA2DPSource.cpp:730] bt_app_av_state_connected(): [BT_AV] bt_app_av_state_connected evt 3
[ 11525][D][BluetoothA2DPSource.cpp:790] bt_app_av_media_proc(): [BT_AV] bt_app_av_media_proc evt 3
[ 11533][I][BluetoothA2DPSource.cpp:813] bt_app_av_media_proc(): [BT_AV] a2dp media start successfully.
[ 11542][D][BluetoothA2DPSource.cpp:301] bt_app_task_handler(): [BT_API] bt_app_task_handler, sig 0x1, 0x1
[ 11552][I][BluetoothA2DPSource.cpp:628] bt_app_av_sm_hdlr(): [BT_AV] bt_app_av_sm_hdlr state 5, evt 0x1
[ 11561][D][BluetoothA2DPSource.cpp:596] process_user_state_callbacks(): [BT_AV] process_user_state_callbacks
[ 11571][D][BluetoothA2DPSource.cpp:615] process_user_state_callbacks(): [BT_AV] process_user_state_callbacks ESP_A2D_AUDIO_STATE_EVT: Started
Bluetooth audio state: Started
[ 11593][D][BluetoothA2DPSource.cpp:730] bt_app_av_state_connected(): [BT_AV] bt_app_av_state_connected evt 1
[ 21215][D][BluetoothA2DPSource.cpp:250] bt_app_work_dispatch(): [BT_API] bt_app_work_dispatch event 0xff00, param len 0
[ 21215][D][BluetoothA2DPSource.cpp:301] bt_app_task_handler(): [BT_API] bt_app_task_handler, sig 0x1, 0xff00
[ 21224][I][BluetoothA2DPSource.cpp:628] bt_app_av_sm_hdlr(): [BT_AV] bt_app_av_sm_hdlr state 5, evt 0xff00
[ 21234][D][BluetoothA2DPSource.cpp:596] process_user_state_callbacks(): [BT_AV] process_user_state_callbacks
[ 21243][D][BluetoothA2DPSource.cpp:730] bt_app_av_state_connected(): [BT_AV] bt_app_av_state_connected evt 65280
[ 21253][D][BluetoothA2DPSource.cpp:790] bt_app_av_media_proc(): [BT_AV] bt_app_av_media_proc evt 65280
[ 31215][D][BluetoothA2DPSource.cpp:250] bt_app_work_dispatch(): [BT_API] bt_app_work_dispatch event 0xff00, param len 0
[ 31215][D][BluetoothA2DPSource.cpp:301] bt_app_task_handler(): [BT_API] bt_app_task_handler, sig 0x1, 0xff00
[ 31224][I][BluetoothA2DPSource.cpp:628] bt_app_av_sm_hdlr(): [BT_AV] bt_app_av_sm_hdlr state 5, evt 0xff00
[ 31234][D][BluetoothA2DPSource.cpp:596] process_user_state_callbacks(): [BT_AV] process_user_state_callbacks
[ 31243][D][BluetoothA2DPSource.cpp:730] bt_app_av_state_connected(): [BT_AV] bt_app_av_state_connected evt 65280
[ 31253][D][BluetoothA2DPSource.cpp:790] bt_app_av_media_proc(): [BT_AV] bt_app_av_media_proc evt 65280




After reading the docs i have to react on ESP_AVRC_RN_PLAY_STATUS_CHANGE.
ESP_AVRC_RN_VOLUME_CHANGE seems to work fine in same way. Am i miss something?

I think almost every bt-source device has a volume control an at least at start/pause button as my kids headset.
Any idea how to go on? Thank's & best regards
Dirk

**Comment by pschatzmann** [comment]
I guess you also need to overwrite the bt_av_hdl_stack_evt to activate some more events: 

```
esp_avrc_rn_evt_bit_mask_operation(ESP_AVRC_BIT_MASK_OP_SET, &evt_set, ESP_AVRC_RN_VOLUME_CHANGE|ESP_AVRC_RN_PLAY_STATUS_CHANGE);
```
After this you should be notified ...

**Reply by sfields-forever** [reply]
I have overwritten the bt_av_hdl_stack_evt by substituting
`esp_avrc_rn_evt_bit_mask_operation(ESP_AVRC_BIT_MASK_OP_SET, &evt_set, ESP_AVRC_RN_VOLUME_CHANGE);`
by
`esp_avrc_rn_evt_bit_mask_operation(ESP_AVRC_BIT_MASK_OP_SET, &evt_set, ESP_AVRC_RN_VOLUME_CHANGE|ESP_AVRC_RN_PLAY_STATUS_CHANGE);	`
but I get the error
`invalid conversion from 'int' to 'esp_avrc_rn_event_ids_t'`
 
What did I do wrong?

Error in more detail:

```
lib/ESP32-A2DP/src/BluetoothA2DPSource.cpp: In member function 'virtual void BluetoothA2DPSource::bt_av_hdl_stack_evt(uint16_t, void*)':
lib/ESP32-A2DP/src/BluetoothA2DPSource.cpp:581:65: error: invalid conversion from 'int' to 'esp_avrc_rn_event_ids_t' [-fpermissive]
                                        ESP_AVRC_RN_VOLUME_CHANGE|ESP_AVRC_RN_PLAY_STATUS_CHANGE);
                                        ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from lib/ESP32-A2DP/src/BluetoothA2DPCommon.h:58,
                 from lib/ESP32-A2DP/src/BluetoothA2DPSource.h:22,
                 from lib/ESP32-A2DP/src/BluetoothA2DPSource.cpp:16:
/home/usb/.platformio/packages/framework-espidf/components/bt/host/bluedroid/api/include/api/esp_avrc_api.h:717:65: note:   initializing argument 3 of 'bool esp_avrc_rn_evt_bit_mask_operation(esp_avrc_bit_mask_op_t, esp_avrc_rn_evt_cap_mask_t*, esp_avrc_rn_event_ids_t)'
                                         esp_avrc_rn_event_ids_t event_id);
                                         ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~
```

**Reply by pschatzmann** [reply]
You forgot to cast!

**Reply by sfields-forever** [reply]
Thankyou. I now have replaced
`esp_avrc_rn_evt_bit_mask_operation(ESP_AVRC_BIT_MASK_OP_SET, &evt_set, ESP_AVRC_RN_VOLUME_CHANGE);`
by
`esp_avrc_rn_evt_bit_mask_operation(ESP_AVRC_BIT_MASK_OP_SET, &evt_set, (esp_avrc_rn_event_ids_t)(ESP_AVRC_RN_VOLUME_CHANGE | ESP_AVRC_RN_PLAY_STATUS_CHANGE));`
and it compiles. 
The log level is set to verbose/debug in addition. But I am not notified when I click on the bt-headphone play/pause/... I get only the sort of messages, @tueddy  reports [here](https://github.com/pschatzmann/ESP32-A2DP/discussions/299#discussioncomment-3694481).
What am I missing? Has anyone managed to be notified by play/pause events?

**Reply by sfields-forever** [reply]
I have made another attempt to capture and display PLAY/PAUSE/NEXT/PREVIOUS events, despite my very limited knowledge. Unfortunately, I haven't been successful. I've made changes to both BluetoothA2DPSource.cpp and BluetoothA2DPSource.h files in an effort to achieve this. 
In my opinion, controlling PLAY/PAUSE/NEXT/PREVIOUS via the Bluetooth device is even more crucial than VOLUME UP/DOWN functionality, and should be an essential component of any Bluetooth library. 
Could someone please assist in making this possible? I've attached the changes I've made to both files. Any help would be greatly appreciated. Thank you in advance for your support!
[BluetoothA2DPSource.cpp.txt](https://github.com/user-attachments/files/18283110/BluetoothA2DPSource.cpp.txt)
[BluetoothA2DPSource.h.txt](https://github.com/user-attachments/files/18283111/BluetoothA2DPSource.h.txt)
[Replace .txt with .patch (otherwise gihubt would not upload the file)]

**Reply by pschatzmann** [reply]
Are you sure that such a thing exists: I have never seen any Bluetooth Speaker which would have PLAY/PAUSE/NEXT/PREVIOUS buttons.

I would rather think that this should be part of your A2DP source and here you can just use the AudioTools Player which provides exactly this functionality.

**Reply by jamesbaber** [reply]
Maybe not Bluetooth speakers but definitely headphones. A lot of them support play/pause, forward, back etc.

**Reply by pschatzmann** [reply]
So if you have such a device, you are in a good position to contribute to this project!

**Reply by sfields-forever** [reply]
> Are you sure that such a thing exists: I have never seen any Bluetooth Speaker which would have PLAY/PAUSE/NEXT/PREVIOUS buttons.

Yes, definitely! I have a Bluetooth device and would argue that practically every Bluetooth over-ear or in-ear headphone has this functionality. Usually, it’s provided by a single button, and depending on how many times you press it, different actions are executed, for example:

    Single press: toggle play/pause
    Double press: next track
    Triple press: previous track

With in-ear headphones, there’s often no button; you just tap on the earpiece. Even my very old Bluetooth speaker (not a headphone) has such a button. In fact, all the Bluetooth devices I own have this feature.


> So if you have such a device, you are in a good position to contribute to this project!

Yes, this is why I really tried.  However, I unfortunately have so little knowledge about this that, despite my sincere efforts, I haven't made any progress, and the code changes above haven't helped. I really wanted to make my son happy with this; it's not for me at all. If you’d like, I would be delighted to send you a Bluetooth device with this functionality. Just let me know your address, and I’d be more than happy to do so!

**Reply by pschatzmann** [reply]
Are the volume buttons currently working ?

**Reply by sfields-forever** [reply]
I was quite sure that they worked in the past. At the moment I only get the messages in the log that the button is pressed, volume does not change (but this could be my fault, I have to recheck). But you get the messages in the log, which is not the case for the play/pause/next buttons (which this discussion is all about).

Logs for volume down:

```
[ 35811][D][BluetoothA2DPSource.cpp:924] bt_app_rc_ct_cb(): [RCCT] bt_app_rc_ct_cb evt 4
[ 35819][D][BluetoothA2DPSource.cpp:258] bt_app_work_dispatch(): [BT_API] bt_app_work_dispatch event 0x4, param len 12
[ 35830][D][BluetoothA2DPSource.cpp:308] bt_app_task_handler(): [BT_API] bt_app_task_handler, signal: 0x1, event: 0x4
[ 35840][D][BluetoothA2DPSource.cpp:1010] bt_av_hdl_avrc_ct_evt(): [RCCT] bt_av_hdl_avrc_ct_evt evt 4
[ 35849][I][BluetoothA2DPSource.cpp:1056] bt_av_hdl_avrc_ct_evt(): [RCCT] AVRC event notification: 13
[ 35858][D][BluetoothA2DPSource.cpp:969] bt_av_notify_evt_handler(): [RCCT] bt_av_notify_evt_handler evt 13
[ 35868][I][BluetoothA2DPSource.cpp:972] bt_av_notify_evt_handler(): [RCCT] Volume changed: 121
[ 35877][I][BluetoothA2DPSource.cpp:975] bt_av_notify_evt_handler(): [RCCT] Set absolute volume: volume 121
[ 35886][I][BluetoothA2DPCommon.h:166] set_volume(): [BT_AV] set_volume: 121
[ 35906][D][BluetoothA2DPSource.cpp:924] bt_app_rc_ct_cb(): [RCCT] bt_app_rc_ct_cb evt 7
[ 35915][D][BluetoothA2DPSource.cpp:258] bt_app_work_dispatch(): [BT_API] bt_app_work_dispatch event 0x7, param len 12
[ 35926][D][BluetoothA2DPSource.cpp:308] bt_app_task_handler(): [BT_API] bt_app_task_handler, signal: 0x1, event: 0x7
[ 35937][D][BluetoothA2DPSource.cpp:1010] bt_av_hdl_avrc_ct_evt(): [RCCT] bt_av_hdl_avrc_ct_evt evt 7
[ 35946][I][BluetoothA2DPSource.cpp:1085] bt_av_hdl_avrc_ct_evt(): [RCCT] Set absolute volume response: volume 121
[ 35956][I][BluetoothA2DPCommon.h:166] set_volume(): [BT_AV] set_volume: 121
```
Logs for volume up:
```
[ 39385][D][BluetoothA2DPSource.cpp:924] bt_app_rc_ct_cb(): [RCCT] bt_app_rc_ct_cb evt 4
[ 39393][D][BluetoothA2DPSource.cpp:258] bt_app_work_dispatch(): [BT_API] bt_app_work_dispatch event 0x4, param len 12
[ 39403][D][BluetoothA2DPSource.cpp:308] bt_app_task_handler(): [BT_API] bt_app_task_handler, signal: 0x1, event: 0x4
[ 39414][D][BluetoothA2DPSource.cpp:1010] bt_av_hdl_avrc_ct_evt(): [RCCT] bt_av_hdl_avrc_ct_evt evt 4
[ 39423][I][BluetoothA2DPSource.cpp:1056] bt_av_hdl_avrc_ct_evt(): [RCCT] AVRC event notification: 13
[ 39432][D][BluetoothA2DPSource.cpp:969] bt_av_notify_evt_handler(): [RCCT] bt_av_notify_evt_handler evt 13
[ 39442][I][BluetoothA2DPSource.cpp:972] bt_av_notify_evt_handler(): [RCCT] Volume changed: 127
[ 39450][I][BluetoothA2DPSource.cpp:975] bt_av_notify_evt_handler(): [RCCT] Set absolute volume: volume 127
[ 39460][I][BluetoothA2DPCommon.h:166] set_volume(): [BT_AV] set_volume: 127
[ 39473][D][BluetoothA2DPSource.cpp:924] bt_app_rc_ct_cb(): [RCCT] bt_app_rc_ct_cb evt 7
[ 39482][D][BluetoothA2DPSource.cpp:258] bt_app_work_dispatch(): [BT_API] bt_app_work_dispatch event 0x7, param len 12
[ 39492][D][BluetoothA2DPSource.cpp:308] bt_app_task_handler(): [BT_API] bt_app_task_handler, signal: 0x1, event: 0x7
[ 39503][D][BluetoothA2DPSource.cpp:1010] bt_av_hdl_avrc_ct_evt(): [RCCT] bt_av_hdl_avrc_ct_evt evt 7
[ 39512][I][BluetoothA2DPSource.cpp:1085] bt_av_hdl_avrc_ct_evt(): [RCCT] Set absolute volume response: volume 127
[ 39522][I][BluetoothA2DPCommon.h:166] set_volume(): [BT_AV] set_volume: 127
```

**Comment by ChadMatsalla** [comment]
Hi Phil!

I created a subclass and I'm trying to detect any AVRC event whatsoever. I see nothing. I've tried two devices and no change (including volume) is ever detected.

Could you give any hint as to the best place and method to log any AVRC event?

I see that you call `esp_avrc_ct_init`, then you register the callback `esp_avrc_ct_register_callback(ccall_bt_app_rc_ct_cb);`. I'm confused as to why `esp_avrc_ct_register_callback(ccall_bt_app_rc_ct_cb)` then apparently delegates to `bt_app_rc_ct_cb`. Am I correct that this line in `bt_app_rc_ct_cb` should log any event?
```
851     ESP_LOGD(BT_RC_CT_TAG, "%s evt %d", __func__, event);
```
If that's not the place, where would be the best place to catch any event?

Thank you,

Chad

**Reply by jamesbaber** [reply]
Hi Chad, did you have any success with this? Similarly, debug log level does not show any AVRC events for me. Thanks

**Reply by pschatzmann** [reply]
The current code only registers ESP_AVRC_RN_VOLUME_CHANGE and you might need to adapt the code to get more

**Reply by jamesbaber** [reply]
Should the ESP_AVRC_RN_VOLUME_CHANGE notifications be logged in debug mode?
I've got ESP IDF version 4.4 but don't get any of the volume change logs I think bt_av_notify_evt_handler should provide.

**Reply by pschatzmann** [reply]
Do you have a bluetooth speaker with volume buttons ? 
You should get the notifications when you change the volume on your bluetooth speaker.

**Reply by ChadMatsalla** [reply]
I'm sorry but I shelved this project. I was trying to turn a bluetooth remote into a general-purpose remote. It was for a lie-detector for my son's 7-th grade class. A kid would sit at it and Alex would decide whether the kid was lying and trigger the correct light by pressing one of the buttons. At the moment I no longer have use for such a thing.

**Reply by jamesbaber** [reply]
Thanks Chad. That sounds like it would have been a fun project!

Phil, I've taken audio-tools' streams-generator-a2dp.ino, changed the log level to Debug or Info but when I change the volume on my Sony WH-1000XM4s I do not get any additional serial logs.
out.source().get_volume() does report a changing volume value which is great. But no messages. I've just updated my audio-tools and esp32-a2dp libraries. I suppose this gives me something to work on.

Thanks!

**Reply by jamesbaber** [reply]
I've found part of my issue. I was setting the AudioLogger log level to Debug, instead of the ESP32 core debug level. I do now get headphone volume level changes showing up in the logs.

**Comment by pschatzmann** [comment]
Just set the log level to Debug and trigger some commands and you should be able to see what methods are called.
I would expect that it is av_hdl_stack_evt

**Comment by franzSusemichel** [comment]
> Hi,
> 
> i have a connected a kids headphone and can play audio from SD successfully. This is my minimal sample: https://github.com/tueddy/BTSourceSD.git
> 
> Now the headphone has a button for play/pause. Is it possible to get an event so that i can react on it for pausing/playing audio? I know the other way around using BT-sink and AVRC events which work fine. I wonder if AVRC events for BT-sink only?
> 
> Found nothing about in wiki/issues Thanks in advance Dirk

Hi, 

not sure if i have the same question, but i'm using a teensy to send start/stop which works fine for me. Unfortunatley i don't have information if play/pause is toggled on my mobile phone (want to show that information on teensy) 

Any recommendations?

**Comment by pschatzmann** [comment, marked as answer]
I extended the functionality to support the buttons in the A2DPSource scenario and added an [corresponding example](https://github.com/pschatzmann/ESP32-A2DP/blob/main/examples/bt_music_sender_avrc/bt_music_sender_avrc.ino) that shows how to handle button presses 

It took me quite some effort and I needed to purchase a new bluetooth speaker with buttons to test this!

**Reply by sfields-forever** [reply]
I thank you most sincerely, both on behalf of my son and numerous others who have been eagerly anticipating this functionality, not only here but in other forums as well. My heartfelt gratitude to you; I have tested it briefly with two different Bluetooth speakers, and the play/pause/next/previous button presses are now being displayed. Thankyou!
